tests/testDCSAM.cpp

```diff
@@ -1344,6 +1344,212 @@ TEST(TestSuite, dcMaxMixture_semantic_slam) {
+    SemanticBearingRangeFactor<gtsam::Pose2, gtsam::Point2> sbr2(
+                xi, l2, lm2_class, semantic_meas, bearing1, range1, br_noise);
+    DCMaxMixtureFactor<SemanticBearingRangeFactor<gtsam::Pose2,
+                                                  gtsam::Point2>> dcmmf(
+                            {xi, l1, l2}, dks, {sbr1, sbr2}, {.5, .5}, false);
+
+    hfg.push_dc(dcmmf);
+    odom = odom * meas;
+    initialGuess.insert(xj, odom);
+    dcsam.update(hfg, initialGuess);
+    DCValues dcvals = dcsam.calculateEstimate();
+
+    size_t mpeClassL1 = dcvals.discrete.at(lc1);
+
+    // Plot poses and landmarks
+#ifdef ENABLE_PLOTTING
+    std::vector<double> xs, ys;
+    for (size_t j = 0; j < i + 2; j++) {
+      xs.push_back(
+          dcvals.continuous.at<gtsam::Pose2>(gtsam::Symbol('x', j)).x());
+      ys.push_back(
+          dcvals.continuous.at<gtsam::Pose2>(gtsam::Symbol('x', j)).y());
+    }
+
+    std::vector<double> lmxs, lmys;
+    lmxs.push_back(
+        dcvals.continuous.at<gtsam::Point2>(gtsam::Symbol('l', 1)).x());
+    lmys.push_back(
+        dcvals.continuous.at<gtsam::Point2>(gtsam::Symbol('l', 1)).y());
+
+    string color = (mpeClassL1 == 0) ? "b" : "orange";
+
+    plt::plot(xs, ys);
+    plt::scatter(lmxs, lmys, {{"color", color}});
+    plt::show();
+#endif
+
+    hfg.clear();
+    initialGuess.clear();
+  }
+
+  gtsam::Symbol x7('x', 7);
+  gtsam::BetweenFactor<gtsam::Pose2> bw(x0, x7, dx * noise, meas_noise);
+
+  hfg.push_nonlinear(bw);
+  dcsam.update(hfg, initialGuess);
+
+  DCValues dcvals = dcsam.calculateEstimate();
+
+  size_t mpeClassL1 = dcvals.discrete.at(lc1);
+
+  // Plot the poses and landmarks
+#ifdef ENABLE_PLOTTING
+  std::vector<double> xs, ys;
+  for (size_t i = 0; i < 8; i++) {
+    xs.push_back(dcvals.continuous.at<gtsam::Pose2>(gtsam::Symbol('x', i)).x());
+    ys.push_back(dcvals.continuous.at<gtsam::Pose2>(gtsam::Symbol('x', i)).y());
+  }
+
+  std::vector<double> lmxs, lmys;
+  lmxs.push_back(
+      dcvals.continuous.at<gtsam::Point2>(gtsam::Symbol('l', 1)).x());
+  lmys.push_back(
+      dcvals.continuous.at<gtsam::Point2>(gtsam::Symbol('l', 1)).y());
+
+  string color = (mpeClassL1 == 0) ? "b" : "orange";
+
+  plt::plot(xs, ys);
+  plt::scatter(lmxs, lmys, {{"color", color}});
+  plt::show();
+#endif
+
+  EXPECT_EQ(mpeClassL1, 1);
+  
+
+  // So far, this same as earlier example. Now let's start removing factors  
+  // and see what happens. 
+  EXPECT_EQ(dcsam.getDiscreteFactorGraph().size(), 18);
+  EXPECT_EQ(dcsam.getNonlinearFactorGraph().size(), 18);
+
+  hfg.clear();
+  initialGuess.clear();
+  initialGuessDiscrete.clear();
+  std::vector<size_t> discreteRemovals{17};
+  gtsam::FactorIndices removals{17};
+
+  dcsam.update(hfg, initialGuess, initialGuessDiscrete, removals, discreteRemovals);
+  
+  EXPECT_EQ(dcsam.getDiscreteFactorGraph().at(17), nullptr);
+  EXPECT_EQ(dcsam.getNonlinearFactorGraph().at(17), nullptr);
```

Comment:
This test doesn't validate the new estimate after the factor removal.

Suggestion: Try to solve a (simple) factor graph with an known solution (that we can verify to within `tol`). Then, remove one or more factors, verify with `EXPECT_EQ` that they've been removed, retrieve the new estimate via `dcsam_.calculateEstimate()` and validate that the new estimate is correct (to within `tol`).

There are also a couple "edge cases" we might want to check on, e.g. what happens if we have a discrete variable with a single discrete factor attached to it, and then we remove that factor? I'm not actually sure what the behavior would be in this case, but it seems like it would be good to know.

Reply:
> what happens if we have a discrete variable with a single discrete factor attached to it, and then we remove that factor?

The variable will be removed (I believe after the `gtsam::VariableIndex` is created, and any subsequent attempts to retrieve information about the variable will throw errors.